src/lib/Libifl/pbsD_deljoblist.c

```diff
@@ -46,160 +46,554 @@
+ * @param[in] extend - extend string for req
+ *
+ * @return struct batch_deljob_status *
+ * @retval list of error objects from server (or NULL)
+ */
+static struct batch_deljob_status *
+deljob_to_server(int fd, int function, char **jobids, int numids, char *extend)
+{
+	if (send_deljob(fd, function, jobids, numids, extend))
 		return NULL;
 
-	return rbsp;
+	return recv_deljob(fd);
 }
 
 /**
- * @brief
- *      -encode a Delete Job List Batch Request
+ * @brief	Broadcast deljob request to other servers in a multi-server set up
+ * 			This handles the cases where a job was moved to another server
  *
- * @par Functionality:
- *              This request is used for delete operation.
- *              
+ * @param[in] c - connection handle (virtual)
+ * @param[in] function - req type
+ * @param[in] jobids - the job id list
+ * @param[in] numjids - the count of job ids
+ * @param[in] extend - extend string for req
  *
- * @param[in] c - socket descriptor
- * @param[in] command - command type
- * @param[in] objtype - object type
- * @param[in] objname - object name
- * @param[in] aoplp - pointer to attropl structure(list)
- * @param[in] prot - PROT_TCP or PROT_TPP
- * @param[in] msgid - message id
+ * @return struct batch_deljob_status *
+ * @retval list of jobs not found on any server
+ */
+static struct batch_deljob_status *
+broadcast_deljob(int c, int function, char **jobids, int numids, char *extend)
+{
+	int *skip_list;
+	svr_conn_t **conns = get_conn_svr_instances(c);
+	int num_svrs = get_num_servers();
+	int i;
+	struct batch_deljob_status **svr_retlist = NULL;
+	struct batch_deljob_status *retlist = NULL;
+	struct batch_deljob_status *iter_list = NULL;
+	struct batch_deljob_status *prev = NULL;
+	struct batch_deljob_status *next = NULL;
+
+	if (conns == NULL) {
+		pbs_errno = PBSE_NOCONNECTS;
+		return NULL;
+	}
+
+	skip_list = calloc(num_svrs, sizeof(int));
+	if (skip_list == NULL) {
+		pbs_errno = PBSE_SYSTEM;
+		return NULL;
+	}
+
+	svr_retlist = calloc(num_svrs, sizeof(struct batch_deljob_status *));
+	if (svr_retlist == NULL) {
+		free(skip_list);
+		pbs_errno = PBSE_SYSTEM;
+		return NULL;
+	}
+
+	/* Broadcast the jids to all servers */
+	for (i = 0; i < num_svrs; i++) {
+		if (send_deljob(conns[i]->sd, function, jobids, numids, extend))
+			skip_list[i] = 1;
+	}
+	for (i = 0; i < num_svrs; i++) {
+		if (!skip_list[i]) {
+			struct batch_deljob_status *ret = NULL;
+
+			ret = recv_deljob(conns[i]->sd);
+			svr_retlist[i] = ret;
+		}
+	}
+
+	/* The list of ids we return has to be the intersection of the lists from all servers
+	 * Use the first server's list as the intersection to be returned
+	 */
+	for (iter_list = svr_retlist[0]; iter_list != NULL; prev = iter_list, iter_list = next) {
+		int job_count = 1;
+		struct batch_deljob_status *iter_list2;
+		next = iter_list->next;
+		int error_code = PBSE_UNKJOBID;
+
+		/* Make sure that this jid is in ALL of the other lists as well */
+		for (i = 1; i < num_svrs; i++) {
```

Comment:
This is a bad assumption to make. pbs_deljoblist can be used to delete jobs from a remote cluster. Right now pbs_connect is not enhanced to talk to a remote msvr-cluster. But the moment we do that we have to revisit all these codes. Because num_svrs is local and not remote.
Instead if you had relied on the size of the connection array, all these will still be valid. But its up to you.

Reply:
If a remote cluster's servers are part of PSI then num_svrs will include them. If not, then qdel does pbs_connect multiple times anyways, once for each cluster, and calls deljoblist once per cluster, so it should not be a problem.

Reply:
on second thought, you are right, in its current form it get_num_servers() won't return the correct value for such a case, so I'll change it, thanks

Reply:
ok, my whole broadcast logic depends on get_num_servers() returning the correct number of servers for the cluster. If and when we start supporting multi-cluster multi-server, we'll have to change it to take a connection descriptor and return the number of servers in the cluster, or add a new function to do so. So, I'll leave the code as it is for now, I don't want to make it confusing by using num_svrs in one for loop and conns[i] != NULL in another.

Reply:
You dont have to use it inconsistently. You could replace all usages of your num_svrs check with conns[i] != NULL . num_svrs check is just the local version of the conns[i] != NULL .